service/history/api/updateworkflow/api.go

```diff
@@ -61,177 +61,204 @@ const (
 	failUpdateWorkflowTaskAttemptCount = 3
 )
 
-func Invoke(
-	ctx context.Context,
-	req *historyservice.UpdateWorkflowExecutionRequest,
+type Updater struct {
+	shardCtx                   shard.Context
+	workflowConsistencyChecker api.WorkflowConsistencyChecker
+	matchingClient             matchingservice.MatchingServiceClient
+	req                        *historyservice.UpdateWorkflowExecutionRequest
+	namespaceID                namespace.ID
+
+	// Variables from workflow.Context and mutable state.
+	// NOTE: They *have to* be copies to avoid data races when using them outside the workflow lease.
```

Comment:
Not all of them. `wfKey`, `upd`, and `directive` can be used outside of lease. `directive` is very tricky though.

Reply:
Better safe than sorry? :)